one_fm/utils.py

```diff
@@ -696,31 +696,59 @@ def get_salary(employee):
 
     return salary_amount
 
-@frappe.whitelist(allow_guest=True)
+@frappe.whitelist()
 def hooked_leave_allocation_builder():
-    # Get Active Employee List
+    '''
+        Function used to create leave allocations
+         - Create Leave Allocation for Employee, who is having a valid leave policy
+        Triggered from hooks as daily event
+    '''
+    # Get Active Employee List and set background job to create leave allocations based on the leave policy
     employee_list = frappe.get_all("Employee", filters={"status": "Active"},
         fields=["name", "date_of_joining", "went_to_hajj", "grade", "leave_policy"])
     frappe.enqueue(leave_allocation_builder, timeout=600, employee_list=employee_list)
 
 def leave_allocation_builder(employee_list):
+    '''
+        Function used to create leave allocations for a given employee list
+         - Create Leave Allocation for Employee, who is having a valid leave policy
+        args:
+            employee_list: List of Employees with minimum data (name, date_of_joining, went_to_hajj, grade and leave_policy)
+    '''
     from erpnext.hr.doctype.leave_allocation.leave_allocation import get_leave_allocation_for_period
-    # Get Leave Policy for employee
+
+    # Get Leave Type details (configurations)
     leave_type_details = get_leave_type_details()
+
+    # Iterate Employee List for finidng employee leave policy to create leave allocation
     for employee in employee_list:
+        # Get employee Leave Policy Object
         leave_policy = get_employee_leave_policy(employee)
+
+        # Check leave policy and details exists
         if leave_policy and leave_policy.leave_policy_details:
+
+            # Iterate Leave policy details to check if there any Leave Allocation exists, if not then create leave allocation
             for policy_detail in leave_policy.leave_policy_details:
+                # Find from_date and to_date to get leave allocation for the current period
                 if getdate(add_years(employee.date_of_joining, 1)) > getdate(nowdate()):
+                    '''
+                        If date of joining + 1 year greater than today, then from_date will be joining date.
+                    '''
                     from_date = getdate(employee.date_of_joining)
                 else:
+                    '''
+                        Else, from_date will be sum of joining date and difference in years
+                            between the date of joining and today
+                    '''
                     datetime1 = get_datetime(getdate(employee.date_of_joining))
                     datetime2 = get_datetime(getdate(nowdate()))
 
                     time_difference = relativedelta(datetime2, datetime1)
                     difference_in_years = time_difference.years
```

Comment:
@pjamsheer what if the `difference_in_years` value is zero that means `from_date` will be `employee.date_of_joining`?

Reply:
employee.date_of_joining itself

Reply:
👍